adding upsertWithWhere

mountain1234585 opened the pull request:
Requires https://github.com/strongloop/loopback-datasource-juggler/pull/1001

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

mountain1234585 commented:
We are still working on adding tests to this PR.

mountain1234585 commented:
Changes to test files have been made for upsertWithWhere. Pls review.

bajtos commented on lib/persisted-model.js:
```diff
@@ -651,6 +673,22 @@ module.exports = function(registry) {
+          description: 'An object of model property name/value pairs' },
+      ],
+      returns: { arg: 'data', type: typeName, root: true },
+      http: { verb: 'put', path: '/upsertWithWhere' },
```
I think we usually use `POST` for non-root endpoints like `/upsertWithWhere`.

Alternatively, we can use `PATCH` to be consistent with `upsert` that's mapped to `PATCH /` by default since LB 3.0.

Amir-61 replied:
@bajtos Could you please confirm me that since this is a new method which is being introduced we don't need to make it configurable like the way we did for `replaceOrCreate`/`updateOrCreate` and `replaceById`/`patchAttributes`, right?

Amir-61 replied:
@bajtos ping Please see ^ Also please let @mountain1234585 know whether they should go for `POST /updateOrCreate` or `PATCH`? I have no strong opinion in this regard.

CC: @raymondfeng

bajtos replied:
>  Could you please confirm me that since this is a new method which is being introduced we don't need to make it configurable like the way we did with replaceOnPUT flag for replaceOrCreate/updateOrCreate and replaceById/patchAttributes, right?

Yes, no need to make this configurable.

> whether they should go for POST /updateOrCreate or PATCH /updateOrCreate? I have no strong opinion in this regard.

Let's go with `POST /upsertWithWhere` or even `POST /updateOrCreateWithWhere`.

Amir-61 replied:
@mountain1234585 Lets got for `POST /upsertWithWhere` and please make sure you add enough test coverage; please look into [PR#2316](https://github.com/strongloop/loopback/pull/2316) which gives you an idea what test coverage should be added in this patch.

bajtos commented:
@slnode ok to test

bajtos commented:
@mountain1234585 thank you for the pull request, it looks mostly good to me. I'll let @Amir-61 to lead the review and give this the final approval.

Amir-61 commented:
I guess this patch is missing lots of test cases; at least tests `test/access-control.integration.js`, `test/model.test.js`, `test/remoting.integration.js`. I recommend if you look into [PR#2316](https://github.com/strongloop/loopback/pull/2316), that should help a lot and give you a good direction.

Amir-61 commented:
Also I see some [commit-linter issues](https://ci.sl.strongloop.com/job/strongloop-commit-linter-jobs/job/loopback/89//console):
```
**  4 problems found:
**    5574dfe - Merge remote-tracking branch 'strongloop/master' in: First line should be 50 characters or less (saw 69)
**    60463b8 - Merge remote-tracking branch 'strongloop/master' in: First line should be 50 characters or less (saw 69)
**    820f78b - Merge remote-tracking branch 'strongloop/master' in: First line should be 50 characters or less (saw 69)
**    51f2b5d - Merge remote-tracking branch 'strongloop/master' in: First line should be 50 characters or less (saw 69)
```
Also you PR needs to be rebased on top of master.

Last but not least I see all of tests fail. Please look into failures and make sure they all pass.

Thanks

mountain1234585 commented:
@Amir-61  While I was running the tests in model.test.js and remoting.integration.js as is without adding any additional tests, there were many failures I could see.  I am running mocha test/remoting.integration.js and mocha test/model.test.js in loopback directory. Please let me know , if I am missing out on configuring . I could do a clean run for access-control.integration.js.

mountain1234585 commented:
Tests are present  in model.test.js, access-control.integration.js and replication.test.js.

Amir-61 commented:
>@Amir-61 While I was running the tests in model.test.js and remoting.integration.js as is without adding any additional tests, there were many failures I could see. I am running mocha test/remoting.integration.js and mocha test/model.test.js in loopback directory. Please let me know , if I am missing out on configuring

You need to rebase all your commits on top of master and I believe the failures have been solved in the latest commits

mountain1234585 commented:
@Amir-61 I have squashed all commits into one and rebased on top of master.

Amir-61 commented:
@rmg I'm looking into failures; however the logs does not give that much info why tests fail. One of the log is [here](http://ci.strongloop.com/job/loopback/node=0.10,os=amazon/4467/consoleFull). Can you find the reason why tests fail which may help @mountain1234585 to resolve it and get this patch pass our CI?

superkhau commented on lib/persisted-model.js:
```diff
@@ -119,6 +119,28 @@ module.exports = function(registry) {
+   * @param {Object} model Updated model instance.
+   */
+
+  PersistedModel.upsertWithWhere  = PersistedModel.patchOrCreateWithWhere =
```
remove one extra space in front of =

superkhau replied:
You can also align them together like:

```
PersistedModel.patchOrCreateWithWhere =
PersistedModel.upsertWithWhere = function upsertWithWhere(...) {
});

mountain1234585 replied:
done

superkhau commented on lib/persisted-model.js:
```diff
@@ -654,6 +676,22 @@ module.exports = function(registry) {
+      aliases: ['patchOrCreateWithWhere'],
+      description: 'Update an existing model instance or ' +
+      'insert a new one into the data source based on the ' +
+      'where criteria.',
```
indent

mountain1234585 replied:
done

Amir-61 replied:
>done
Although nitpick, @superkhau's concern is not addressed.

bajtos replied:
+1, please fix the indentation.

@superkhau @Amir-61 I am surprised `eslint` did not catch this problem, would you mind taking a look at our eslint config and see if we are missing a directive?

superkhau replied:
Yes weird, should be automated. Upon a quick look in our configs, it is set properly. @mountain1234585 Did you run the linter/tests before recommiting the PR?

mountain1234585 replied:
@superkhau I did run the eslint before recommitting and it dint report me any errors. I have changed the indentation now. Pls check if that's fine.

superkhau commented on test/model.test.js:
```diff
@@ -146,6 +146,37 @@ describe.onServer('Remote Methods', function() {
 
+  describe('Model.upsertWithWhere(where, data, callback)', function() {
+    it('Updates when a Model instance is retreived from data source', function(done) {
+      (new TaskEmitter())
```
```
var taskEmitter = new TaskEmitter();
taskEmitter
  .task...
  .task...

mountain1234585 replied:
done

superkhau commented on test/model.test.js:
```diff
@@ -146,6 +146,37 @@ describe.onServer('Remote Methods', function() {
+    });
+
+    it('Creates when no Model instance is retreived from data source', function(done) {
+      (new TaskEmitter())
```
ditto

mountain1234585 replied:
done

superkhau commented on test/model.test.js:
```diff
@@ -146,6 +146,37 @@ describe.onServer('Remote Methods', function() {
+          .on('done', function() {
+            User.upsertWithWhere({ second: 'pill' }, { second: 'jones' }, function(err, instances) {
+              var id = instances.id;
+              User.findById(id, function(err, instance) {
```
`if (err) return done(err)`

superkhau replied:
`user` instead of `instance`

mountain1234585 replied:
done

superkhau commented on test/model.test.js:
```diff
@@ -146,6 +146,37 @@ describe.onServer('Remote Methods', function() {
+      (new TaskEmitter())
+          .task(User, 'create', { first: 'simon', second: 'somers' })
+          .on('done', function() {
+            User.upsertWithWhere({ first: 'somers' }, { first: 'Simon' }, function(err, instances) {
```
`if (err) return done(err)`

mountain1234585 replied:
done

superkhau commented on test/model.test.js:
```diff
@@ -146,6 +146,37 @@ describe.onServer('Remote Methods', function() {
+          .on('done', function() {
+            User.upsertWithWhere({ first: 'somers' }, { first: 'Simon' }, function(err, instances) {
+              var id = instances.id;
+              User.findById(id, function(err, instance) {
```
`if (err) return done(err)`

superkhau replied:
also we should use `user` instead of `instance`

mountain1234585 replied:
done

superkhau commented on test/model.test.js:
```diff
@@ -146,6 +146,37 @@ describe.onServer('Remote Methods', function() {
+          .task(User, 'create', { first: 'jill', second: 'pill' })
+          .task(User, 'create', { first: 'bob', second: 'sob' })
+          .on('done', function() {
+            User.upsertWithWhere({ second: 'pill' }, { second: 'jones' }, function(err, instances) {
```
`users` instead of `instances`

superkhau replied:
`if (err) return done(err)`

mountain1234585 replied:
done

superkhau commented:
@Amir-61 Reviewed, mostly LGTM except the comments I left. I'll let you take it from here.

mountain1234585 commented:
@superkhau I have addressed all your comments now.
@Amir-61 : I have changed the http verb for upsertWithWhere to POST from PUT.  Also have modified that in the tests. Squashed the commits and rebased on top of master.

Amir-61 commented:
@rmg Could you help @mountain1234585 to figure out what is the failure reason on CI? I had a quick look at logs, but I could not find the reason why they fail.

Amir-61 commented:
Please see CI failures. Also please make sure we have the same test coverage as [PR#2316](https://github.com/strongloop/loopback/pull/2316).

mountain1234585 commented:
@Amir-61  .. I had a look at PR#2316 and the tests are present in test/model.test.js, test/access-control.integration.js, test/remote.integration.js , test/replication.test.js and test/data-source.test.js. 
Please let me know, if there is anything specific you are looking at?

Amir-61 commented:
@slnode test please

Amir-61 commented:
>I saw this error while running my tests locally, but replacing the loopback-datasource-juggler module which contains all my changes resolved the issue. I am not sure what needs to be done to do the same process for these CI builds. Could you please help ?

Now that your PR#1001 is landed tests start passing;

@rmg I see `NVM` machines all fail. Are they setup or they are not ready yet?

Amir-61 commented:
@mountain1234585 
>I had a look at PR#2316 and the tests are present in test/model.test.js, test/access-control.integration.js, test/remote.integration.js , test/replication.test.js and test/data-source.test.js. 
Please let me know, if there is anything specific you are looking at?

I had a quick look into this PR and I think it is missing tests for ACL specifically which should be more in `test/access-control.integration.js`; also I think you need to add ACL in `common/models/user.json` as well.

Also, I landed your PR#1001 to master branch which is LB 3.0 (Our future version) and tried to backport to LB 2.x (Our current version); however lots of connectors started failing your added tests... you may look into them to accelerate the release procedure for LB 2.x (our current version). [PR#1052](https://github.com/strongloop/loopback-datasource-juggler/pull/1052) is the one which backports to LB 2.x.

mountain1234585 commented:
@Amir-61 @superkhau 
I have added tests under user and bank model in test/access-control.integration.js. I also added ACL in common/models/user.json, as below : 
{
      "principalType": "ROLE",
      "principalId": "$owner",
      "permission": "ALLOW",
      "property": "upsertWithWhere"
    }
Now, I added test for accountWithReplaceOnPUTtrue and it passed for the below  cases : 
lt.it.shouldBeDeniedWhenCalledAnonymously('POST', '/api/accounts-replacing/upsertWithWhere');
    lt.it.shouldBeDeniedWhenCalledUnauthenticated('POST', '/api/accounts-replacing/upsertWithWhere');
    lt.it.shouldBeDeniedWhenCalledByUser(CURRENT_USER, 'POST', '/api/accounts-replacing/upsertWithWhere');

Now, when I add it for logged in user , it fails with 401, AUTHORIZATION_REQUIRED.  Is there something I am missing out ? Any pointers on this would be great.

mountain1234585 commented:
@Amir-61 and @superkhau 
 I have added the tests applicable for upsertWithWhere and also the ACL in user.json. 
Also, as mentioned in this comment https://github.com/strongloop/loopback/pull/2539#issuecomment-240405759 above, pls let me know if that needs to be added.

Amir-61 commented:
@mountain1234585,

Let my clarify my earlier [comment](https://github.com/strongloop/loopback/pull/2539#issuecomment-240205629); when we tried to backport your [juggler-PR#1001](https://github.com/strongloop/loopback-datasource-juggler/pull/1001), most of the connectors start failing on CI and we are not going to land it to `2.x` until these failures get fixed. [PR#1052](https://github.com/strongloop/loopback-datasource-juggler/pull/1052) is the one which backports your methods to 2.x, which is the current LoopBack version. So before continuing reviewing this PR you may look into that PR and try to solve the failures. @bajtos, @superkhau and myself will be helping you down the road as much as our bandwidth allows, since we have other responsibilities in our team as well; tbh, right now off the top of my head I don't have any strong opinion why they fail in `2.x`; that requires some investigation and probably trying to reproduce the issue on local machines. So please look into that backporting PR and try to reproduce the issue on local machine before going further, we will also post our findings as well, if we find out why your patch breaks lots of test cases.

mountain1234585 commented:
@Amir-61 @superkhau 
 I have updated [PR#1052](https://github.com/strongloop/loopback-datasource-juggler/pull/1052) with my findings.

Amir-61 commented on test/access-control.integration.js:
```diff
@@ -324,4 +332,23 @@ describe('access control - integration', function() {
+        });
+      });
+    });
+    lt.beforeEach.givenModel('accountWithReplaceOnPUTtrue');
```
`accountWithReplaceOnPUTtrue` is added specifically for testing `replace` methods when `replaceOnPUT` flag is set to true; I don't think you should test with that model; I believe you should test it with `account` model instead to make the purpose of test clear.

bajtos replied:
+1, please use `account` instead.

mountain1234585 replied:
@bajtos  Should I create a new model called account under test/fixtures/common/models .   

I see in test/fixtures/common/models, there is already account.json and inside which the "name" is specified as "accountWithReplaceOnPUTtrue" and plural as "accounts-replacing". Please let me know. For now, I have deleted this block, since I have already added tests for "users" model and "banks" model respectively : 

lt.it.shouldBeDeniedWhenCalledAnonymously('POST', '/api/users/upsertWithWhere');
lt.it.shouldBeDeniedWhenCalledUnauthenticated('POST', '/api/users/upsertWithWhere');
lt.it.shouldBeDeniedWhenCalledByUser(CURRENT_USER, 'POST', '/api/users/upsertWithWhere');

lt.it.shouldBeDeniedWhenCalledAnonymously('POST', '/api/banks/upsertWithWhere');
lt.it.shouldBeDeniedWhenCalledUnauthenticated('POST', '/api/banks/upsertWithWhere');
lt.it.shouldBeDeniedWhenCalledByUser(CURRENT_USER, 'POST', '/api/banks/upsertWithWhere');

Amir-61 replied:
>there is already account.json and inside which the "name" is specified as "accountWithReplaceOnPUTtrue" and plural as "accounts-replacing". 

Oh right; by default LoopBack 3.0 sets `replaceOnPUT` to true, whereas in LoopBack 2.x it is set to false by default; so that's why `account.json` has a name property `accountWithReplaceOnPUTtrue`. How about using other available models in `test/fixtures/access-control/common/models`; probably you can use `bank` model for instance. I do not have a strong opinion I leave it to you and @bajtos.

Amir-61 commented:
Regarding your question in [2359-comment](https://github.com/strongloop/loopback/pull/2539#issuecomment-240405759), it seems in the example code you posted you are using `accounts-replacing` model which is added specifically for testing `replace` methods when `replaceOnPUT` flag is set to true; I don't think you should test with that model; I believe you should test it with regular models instead to make the purpose of test clear. Apart from that, you should add the failing test in the actual code of the PR so then we can comment on that.

I'm busy with landing  your 2.x backporting beside other tasks/responsibilities that I have; I sent [PR-1061](https://github.com/strongloop/loopback-datasource-juggler/pull/1061) which implements `notify` for `find` method to make your 2.x PR pass. @bajtos PTAL at this PR. The patch almost looks good to me; however my only concern was the missing test cases.

bajtos commented on test/remoting.integration.js:
```diff
@@ -235,6 +235,32 @@ describe('With model.settings.replaceOnPUT true', function() {
+      'deleteById(id:any):object DELETE /stores/:id',
+      'count(where:object):number GET /stores/count',
+      'prototype.patchAttributes(data:object):store PATCH /stores/:id',
+      'createChangeStream(options:object):ReadableStream POST /stores/change-stream',
```
Please move this test to `describe('Model shared classes', ` (the block which starts on line 76) and assert only on the methods related to your patch.

```js
it('has upsertWithWhere remote method', function() {
  var storeClass = findClass('store');
  var methods = getFormattedMethodsExcludingRelations(storeClass.methods);
  var expectedMethods = [
    'upsertWithWhere(where:object,data:object):store POST /stores/upsertWithWhere',
  ];
  expect(methods).to.include.members(expectedMethods);
});
```

mountain1234585 replied:
done. thanks.

bajtos commented:
I quickly reviewed the patch, the code looks mostly good. As I see it, there are three line comments above to address, and then the missing test for ACL and `lt.it.shouldBeAllowed` to implement/figure out.

mountain1234585 commented:
@bajtos 
I have addressed two comments out of 3 above. For using the account model, I have posted my query in the comment.
Also, I have already added tests in test/access-control.integration.js under /users and /banks. 

I tried adding lt.it.shouldBeAllowed under the block lt.describe.whenLoggedInAsUser(starting at line 96) in the describe('/users', function() {} block. But it failed with 404. 
 lt.describe.whenCalledRemotely('POST', '/api/users/upsertWithWhere', function() {
        lt.it.shouldBeAllowed();
      });
I haven't added this test case , since it dint pass my local setup.
Pls let me know what to do about it.

Amir-61 commented:
> I tried adding lt.it.shouldBeAllowed under the block lt.describe.whenLoggedInAsUser(starting at line 96) in the describe('/users', function() {} block. But it failed with 404. 

It gives 401; You have already added:
```
    {
      "principalType": "ROLE",
      "principalId": "$owner",
      "permission": "ALLOW",
      "property": "upsertWithWhere"
    },
```
TBH I'm not sure why; @bajtos could you please help @mountain1234585 with this?

Thanks!


PS: @mountain1234585 Please when copy&paste code put them in the [code-block](https://help.github.com/articles/creating-and-highlighting-code-blocks/)

Amir-61 commented:
@slnode test please

bajtos commented:
@mountain1234585 please commit your code to github, so that I can run your failing unit test locally on my machine. Also please rebase your patch on top of the current master.

SubuIBM commented:
@bajtos I have added the failing test to my commit now. Please take a look.

Also, I noticed, the following test case is failing in recent build: This however, is not relevant to the upsertWithWhere changes.
```
KeyValueModel REST API provides "keys(filter)" at "GET /keys":
     Error: done() invoked with non-Error: {"statusCode":500,"message":"Internal Server Error"}
    at Test.<anonymous> (test/key-value-model.test.js:132:40)
```

mountain1234585 commented:
@bajtos  I din't realize , I dint log out of Subu's id earlier ( we were trying out something) and posted the above comment :)
cc @SubuIBM 
Please note that, I have not commented out the failing test after adding them. So, the builds will fail too.

mountain1234585 commented:
@bajtos 
I have added the failing test to the recent commit as per your suggestion. Please take a look.

Also , I noticed in the recent builds one test case ( not relevant to upsertWithWhere feature is failing).
```
KeyValueModel REST API provides "keys(filter)" at "GET /keys":
     Error: done() invoked with non-Error: {"statusCode":500,"message":"Internal Server Error"}
    at Test.<anonymous> (test/key-value-model.test.js:132:40)
```

bajtos commented:
> access control - integration /users POST /api/users/upsertWithWhere should be allowed

Oh, I see it now. Upsert does not support `$owner` role. This is a known limitation of `upsert`, see e.g. https://github.com/strongloop/loopback-sdk-angular/issues/125. You can see that there is no ACL entry for `upsert` either.

We should follow `upsert` behaviour for `upsertWithWhere` and simply don't allow `upsertWithWhere` for the `User` model.

We should eventually find a way how to fix `upsert` (and `upsertWithWhere`) for `User` model, but that's out of scope of this pull request.

I am proposing to revert the changes in `common/models/user.json` and remove the appropriate tests.

mountain1234585 commented:
@bajtos  I have made the changes as you have suggested and committed the changes. PTAL.

bajtos commented:
The patch LGTM in general. I'll leave the final approval to @Amir-61, he has better understanding of the overall context of this work.

Amir-61 commented:
@slnode test please

Amir-61 commented:
@slnode test please

Amir-61 commented:
LGTM; however we need to make sure CI tests pass for Loopback and its dependencies

@rmg CI is failing for PhantomJS; any thought about that; please see logs [here](http://ci.strongloop.com/job/loopback/node=0.12,os=windows/4776/console)

Amir-61 commented:
@slnode test please

rmg commented:
@Amir-61 the failing build seems to have been cleared already, but since it is PhantomJS + Windows, I imagine it was probably related to running out of handles again. The VMs were recently rebooted, so that will hopefully give us another few days without those errors coming back.

Amir-61 commented:
Landed. Thanks for the contribution!
